### PR TITLE
Ignore link-local and invalid IP addresses when waiting for IP.

### PR DESCRIPTION
**Pull request by rhencke**
vSphere may provide a link-local IP address when waiting for an IP
address, but this can't be used to connect to the VM.  This especially
seems to happen on Windows VMs.

### CODE REVIEW

File: driver/vm.go [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 	"fmt"
 	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/property"
 	"github.com/vmware/govmomi/vim25/mo"
 	"github.com/vmware/govmomi/vim25/types"
-	"time"
+	"net"
 	"strings"
-	"context"
+	"time"
```
Comment (rhencke) [inline driver/vm.go]: `gofmt` did this.  I can re-arrange to the prior order if desired.

File: driver/vm.go [hunk 2]
```diff
@@ -54,7 +56,7 @@ type CreateConfig struct {
 	Network       string // "" for default network
 	NetworkCard   string // example: vmxnet3
 	USBController bool
-	Version       uint // example: 10
+	Version       uint   // example: 10
```
Comment (rhencke) [inline driver/vm.go]: `gofmt` did this. I can restore if desired.

File: driver/vm.go [hunk 3]
```diff
@@ -337,8 +339,22 @@ func (vm *VirtualMachine) PowerOn() error {
 	return err
 }
 
-func (vm *VirtualMachine) WaitForIP(ctx context.Context) (string, error) {
-	return vm.vm.WaitForIP(ctx)
+func (vm *VirtualMachine) WaitForIP(ctx context.Context) (ip string, err error) {
```
Comment (rhencke) [inline driver/vm.go]: The new code is roughly based on how `vm.vm.WaitForIP(ctx)` works, by watching for changes to `guest.ipAddress`.  Unlike the original function, though, this imposes the additional restrictions that the returned IP address must:
* be valid
* not be [link-local unicast](https://en.wikipedia.org/wiki/Link-local_address#IPv6)

File: driver/vm.go [hunk 4]
```diff
@@ -337,8 +339,22 @@ func (vm *VirtualMachine) PowerOn() error {
...
+	collector := property.DefaultCollector(vm.vm.Client())
+	err = property.Wait(ctx, collector, vm.vm.Reference(), []string{"guest.ipAddress"}, func(pc []types.PropertyChange) bool {
+		for _, c := range pc {
+			if c.Name == "guest.ipAddress" && c.Op == types.PropertyChangeOpAssign && c.Val != nil {
```
Comment (rhencke) [inline driver/vm.go]: Monitor for an assignment to `guest.ipAddress` that is not `nil`.

File: driver/vm.go [hunk 5]
```diff
@@ -337,8 +339,22 @@ func (vm *VirtualMachine) PowerOn() error {
...
+				ipCandidate := net.ParseIP(c.Val.(string))
```
Comment (rhencke) [inline driver/vm.go]: Try to parse the value we received.  If it is not a valid IP address, `ipCandidate` will be `nil`.

File: driver/vm.go [hunk 6]
```diff
@@ -337,8 +339,22 @@ func (vm *VirtualMachine) PowerOn() error {
...
+				if ipCandidate != nil && !ipCandidate.IsLinkLocalUnicast() {
```
Comment (rhencke) [inline driver/vm.go]: Ensure the value received is both a valid IP, as well as not [link-local unicast](https://en.wikipedia.org/wiki/Link-local_address#IPv6).

File: driver/vm.go [hunk 7]
```diff
@@ -337,8 +339,22 @@ func (vm *VirtualMachine) PowerOn() error {
...
+					ip = ipCandidate.String()
+					return true
```
Comment (rhencke) [inline driver/vm.go]: We have found what appears to be a valid IP.  Assign it to `ip`, and stop watching for property changes.

File: driver/vm.go [hunk 8]
```diff
@@ -337,8 +339,22 @@ func (vm *VirtualMachine) PowerOn() error {
...
+				}
+			}
+		}
+		return false
```
Comment (rhencke) [inline driver/vm.go]: Continue watching for property changes - we have not found a valid IP yet.

**Comment by mkuzmin** [comment]
Hi Robert!

VMware Tools 10.2 has several parameters to filter reported addresses: https://github.com/vmware/open-vm-tools/issues/100#issuecomment-401915103

Do you think we still have to fix that on Packer side?

**Comment by rhencke** [comment]
Hi Michael,

Thank you for your review and response! My understanding of open-vm-tools was that it was not available/used for Windows.  However, maybe this same option exists in the official VMWare Tools - I will investigate this.

Thanks, and also thanks for this project - it's been a wonderfully helpful tool.

**Comment by rhencke** [comment]
Hi Michael,

You are correct, and this feature is available in the same version of the official VMWare Tools: 
 https://docs.vmware.com/en/VMware-Tools/10.2/rn/vmware-tools-1021-release-notes.html

I will work to see if this mechanism can be used to solve this issue, instead.  Thank you for the pointer.

**Pull request closed without merging**